wagtail_streamfield_migration_toolkit/utils.py

```diff
@@ -1,6 +1,188 @@
+    altered_value = {}
+    for key in struct_block_value:
+        child_value = struct_block_value[key]
+
+        # If the block is not at the start of `block_path`, then neither it nor it's children are
+        # blocks that we need to change.
+        if not is_block_at_path_start(key, block_path):
+            altered_value[key] = child_value
+
+        else:
+            child_block_def = block_def.child_blocks[key]
+            altered_child_value = map_block_value(
+                child_value,
+                block_def=child_block_def,
+                block_path=block_path[1:],
+                **kwargs
+            )
+            altered_value[key] = altered_child_value
+
+    return altered_value
+
+
+def map_list_block_value(list_block_value, block_def, block_path, **kwargs):
+
+    mapped_value = []
+    for child_block in list_block_value:
+
+        mapped_child_value = map_block_value(
+            child_block["value"],
+            block_def=block_def.child_block,
+            block_path=block_path[1:],
+            **kwargs
+        )
+
+        mapped_value.append({**child_block, "value": mapped_child_value})
+
+    return mapped_value
+
+
+def apply_changes_to_raw_data(
+    raw_data, block_path_str, operation, streamfield, **kwargs
+):
+    """
+    Applies changes to raw stream data
+
+    Args:
+        raw_data:
+            The current stream data (a list of top level blocks)
+        block_path_str:
+            Should be a '.' separated list of names of the blocks from the top level block to
+            the direct parent of the nested block which is to be changed.
+
+            eg:- 'simplestream.struct1' would point to,
+                [..., { type: simplestream, value: [..., { type: struct1, value: {...} }] }]
+
+            NOTE: If we're directly applying changes on the top level stream block, then this will
+            be None.
+
+            NOTE: When the path contains a ListBlock child, 'item' must be added to the block as
+            the name of said child.
+
+            eg:- 'list1.item.stream1' where the list child is a StructBlock would point to,
+                [
+                    ...,
+                    {
+                        type: list1,
+                        value: [
+                            {
+                                type: item,
+                                value: { ..., stream1: [...] }
+                            },
+                            ...
+                        ]
+                    }
+                ]
+        operation:
+            A subclass of `operations.BaseBlockOperation`. It will have the `apply` method
+            for applying changes to the parent block of the block type that must be changed.
+        streamfield:
+            The streamfield for which data is being migrated. This is used to get the definitions
+            of the blocks.
+
+    Returns:
+        altered_raw_data:
+    """
+
+    if block_path_str is not None:
```

Comment:
Maybe we should use an empty string for this, rather than `None`? That feels like it could indicate "top of the streamfield" pretty intuitively too